### DISCUSSION TITLE
API with group permission does not work

### DISCUSSION THREAD

**Discussion by cfelbertp** [requarks/wiki#6216] [Error / Bug Report]
Hi,

I enabled the API access and get the content of a single page, e.g.:

```
{
  pages {
    single (id: 15) {
      path
      title
      content
      createdAt
      updatedAt
    }
  }
}
```

Then I created a new API key and selected a group for permissions. I assigned all available permission in the content section. 

Under page rules, I configured the following:

Allow read:pages, read:source, read:history, read:asset
Path starts with | any locale | /

When posting the request, I receive a 200: 'You are not authorized to view this page.' It does not matter what id I use. I also tried to configure the exact path of pages I tried to fetch. But I get a list of all pages posting:

```
{
  pages {
    list (orderBy: TITLE) {
      id
      path
      title
    }
  }
}
```
I use Wiki.JS 2.5.297 on Cloudron 7.3.6

**Comment by rrg92** [comment]
You give the group same equivalent system level permission?

**Reply by cfelbertp** [reply]
@rrg92, I assigned all available permissions in the content section to this group.

**Comment by cfelbertp** [comment]
The rules/permissions work for the list query, but I cannot access any single page.

**Comment by cfelbertp** [comment]
Does no one have this problem or is there a lack of information to better understand the issue?

**Comment by gcleaves** [comment]
I have the same problem. The API token created with a group cannot view a page even if web users of the group have access to read the page. 

An API token with full permission is able to view the page.

**Comment by cfelbertp** [comment]
Very strange that even after weeks, there is no reaction whether this is a bug and whether it will be fixed.

**Comment by MetallicGoat** [comment]
Having the same difficulties here.

**Comment by MrEAlderson** [comment]
One way to overcome this would be to grant full access when creating the key. But that's not really a proper solution...

**Comment by jprevost-pro** [comment]
I'm experiencing the same issue.
I'm able to list all pages but get a PageViewForbidden 6013 exception

**Comment by l7s** [comment]
Same problem here. I'm able to list all pages but get a PageViewForbidden 6013 exception for query like below:

```
{ pages { single(id: PAGE_ID) { id path title locale updatedAt } }}
```

**Comment by NellsRelo** [comment]
I ran into this as well, but was able to get it working by giving the group Manage Page permissions. This isn't an ideal solution, as the API key could be then used to move or rename pages, but may shed light on the nature of the bug

**Comment by bullno1** [comment]
It explicitly checks for `manage:pages` and `delete:pages'` here: https://github.com/requarks/wiki/blob/main/server/graph/resolvers/page.js#L155

**Reply by lmahoney1** [reply]
Adding the `manage:pages` and `delete:pages` permission to my group wasn't working until I went to the 'page rules' within the group and selected the `manage:pages` and `delete:pages` in the applicable path(s):

<img width="1591" height="936" alt="image" src="https://github.com/user-attachments/assets/4b7a6e30-a97f-4012-a3fe-0520a348c5a4" />

After doing that I'm able to get the content of a single page via an API token properly with the following query

```
{
	pages {
		single(id: 1234) {
			content
		}
	}
}
```

@bullno1 appreciate you finding and pointing this out!

**Comment by nilleb** [comment]
I am admin of the platform yet I get an error 6013.

Given that this is a page query I would not agree about the fact that manage:pages or delete:pages is required.

**Comment by jaredbrogan** [comment]
@NGPixel, are you tracking on this? Having the ability to have a functional read-only permission for API tokens would be a big improvement.

**Comment by pdennis** [comment]
same problem here

**Comment by greg-konv** [comment]
Hi,

Up, the issue is still on going, it requires at least Manage pages in Permissions and Rename / Move Pages in Page Rules:

<img width="459" height="523" alt="image" src="https://github.com/user-attachments/assets/ff0423a0-d500-4a52-8183-99fa4a16f713" />

This is not ideal for an access in read-only mode. Any ideas? Thanks.

**Comment by Meisterzunge** [comment]
# Bug Report: pages.single ignores page rules

## Severity

**High** — This bug causes permission misbehavior that prevents users with valid page-rule-based access from reading page content, while `pages.list` and `pages.search` still expose the pages' existence.

## Environment

- **Wiki.js version:** 6.1.2 (and likely earlier 6.x versions)
- **Database:** PostgreSQL (observed)
- **Authentication:** API key (JWT) with group-assigned permissions

## Bug Summary

`pages.single` requires `manage:pages` in the **group's global permissions array** — it does **not** check page rules. This creates an inconsistency: a user granted read access via a page rule can list pages but cannot view their content.

## Expected Behavior

A user with a valid page rule that grants `read:pages` and/or `write:pages` should be able to use `pages.single` to read page content, just as `pages.list` respects the same page rule.

## Actual Behavior

`pages.single` always checks `manage:pages` in the group's **global permissions array** and ignores page rules entirely.

`server/graph/resolvers/page.js` — `pages.single` resolver:

    if (WIKI.auth.checkAccess(context.req.user, ['manage:pages', 'delete:pages'], {
      path: page.path,
      locale: page.localeCode
    })) {
      return { ...page, locale: page.localeCode, ... }
    } else {
      throw new WIKI.Error.PageViewForbidden()  // ← always thrown for page-rule-only access
    }

By contrast, `pages.list` applies the full permission chain (global permissions **OR** page rules):

    const results = rawResults.filter(r => {
      return WIKI.auth.checkAccess(context.req.user, ['read:pages'], {
        path: r.path,
        locale: r.locale
      })
    })

## Reproduction Steps

1. Create a new Wiki.js group (or use an existing one)
2. Set **global permissions** to: `read:pages`, `write:pages` (no `manage:pages`)
3. Add a **page rule** with:
   - Path: `""` (match all)
   - Roles: `read:pages`, `write:pages`
4. Create an API key assigned to that group
5. Call `pages.list` — returns pages (PASS)
6. Call `pages.single(id: N)` — returns `6013 PageViewForbidden` (FAIL)

## Root Cause

In `server/graph/resolvers/page.js`, the `single` resolver hard-codes `['manage:pages', 'delete:pages']` as the required permission. The `checkAccess` helper does apply page rules, but the `manage:pages` requirement cannot be satisfied by a page rule — it only exists in global permissions.

## Suggested Fix

Change the `pages.single` permission check from `manage:pages` to `read:pages`:

    // Before
    if (WIKI.auth.checkAccess(context.req.user, ['manage:pages', 'delete:pages'], { path: page.path, locale: page.localeCode })) {

    // After
    if (WIKI.auth.checkAccess(context.req.user, ['read:pages', 'manage:pages', 'delete:pages'], { path: page.path, locale: page.localeCode })) {

This allows:
- Users with `read:pages` via page rules → can read page content
- Users with `manage:pages` globally → can read page content (unchanged)
- Users with `delete:pages` globally → can read page content (unchanged, for conflict checking)

## Additional Notes

- `pages.search` likely has the same issue — it also checks `read:pages` via `checkAccess` but returns partial results only, so the inconsistency is less visible.
- `pages.list` already handles this correctly by using `read:pages` (not `manage:pages`) in its permission check, making the `pages.single` behavior a clear outlier.

**Comment by CarlaFernandez** [comment]
Bumping this thread since the behaviour is still there. @nicolasg-rocketleap any way to mark this as an open issue?
